File: ai-service/app/rag/metadata_filters.py
"""
Metadata-aware retrieval filters for coach chat RAG.
Applied during Node hybrid search (before reranking).
"""

from __future__ import annotations

from typing import Any

from app.rag.levels import L1_INTERNAL, L5_BOOKS


def _as_list(value: Any) -> list[str]:
    if not isinstance(value, list):
        return []
    return [str(v).strip() for v in value if v and str(v).strip() and str(v).strip() != "none"]


def build_metadata_filters(
    *,
    intent: str,
    context_bundle: dict[str, Any] | None,
    locale: str = "en",
    level: str | None = None,
) -> dict[str, Any] | None:
    """
    Map intent + CAG bundle (+ optional level) → Node metadataFilters payload.

    | Intent               | Filters                                      |
    |----------------------|----------------------------------------------|
    | exercise_alternative | primaryMuscles, difficulty, injury exclusions|
    | nutrition            | dietType, religiousDiet, allergens           |
    | platform_help        | L1: docType=platform; L5: exclude catalog    |
    """
    bundle = context_bundle or {}
    profile = bundle.get("profile") or {}
    constraints = bundle.get("constraints") or {}
    onboarding = bundle.get("onboardingSummary") or {}
    nutrition_onboarding = (bundle.get("onboardingByFlow") or {}).get("nutrition") or {}

    filters: dict[str, Any] = {
        "chunkRoles": ["child", "standalone"],
        "requireEmbedding": True,
    }

    resolved = str(intent or "general")

    if resolved in ("platform_help", "unclear"):
        if level == L5_BOOKS:
            filters["excludeTags"] = ["catalog"]
            return filters
        if level is None or level == L1_INTERNAL:
            filters["docType"] = "platform"
            filters["excludeTags"] = ["catalog", "books"]
            if locale in ("en", "ar"):
                filters["locale"] = locale
            return filters
        return filters

    if resolved in ("exercise_alternative", "workout"):
        level = str(profile.get("fitnessLevel") or onboarding.get("fitnessLevel") or "").lower()
        if "beginner" in level or "novice" in level:
            filters["difficulty"] = ["beginner"]
            filters["excludeDifficulty"] = ["advanced"]
        elif "advanced" in level or "expert" in level:
            filters["difficulty"] = ["intermediate", "advanced"]
        elif level:
            filters["difficulty"] = ["beginner", "intermediate"]

        injuries = _as_list(constraints.get("injuries"))
        if injuries:
            filters["excludeExerciseNames"] = injuries

        workout_today = bundle.get("workoutToday") or {}
        exercises = workout_today.get("exercises") or workout_today.get("loggedExercises") or []
        if isinstance(exercises, list) and exercises:
            first = exercises[0] if isinstance(exercises[0], dict) else {}
            muscles = first.get("primaryMuscles") or first.get("muscles") or []
            if not muscles and first.get("muscleGroup"):
                muscles = [first["muscleGroup"]]
            if muscles:
                filters["primaryMuscles"] = [str(m) for m in muscles[:4]]
        return filters

    if resolved == "nutrition":
        diet_type = (
            constraints.get("dietType")
            or onboarding.get("dietType")
            or nutrition_onboarding.get("dietType")
        )
        if diet_type:
            filters["dietType"] = str(diet_type)

        religious = constraints.get("religiousDiet") or onboarding.get("religiousDiet")
        rel_list = _as_list(religious if isinstance(religious, list) else [religious] if religious else [])
        dietary = next(
            (r for r in rel_list if r.lower() not in ("none", "ramadan", "christian_fasting")),
            rel_list[0] if rel_list else "",
        )
        if dietary:
            filters["religiousDiet"] = str(dietary)

        allergy_filters = constraints.get("allergyFilters") or {}
        if isinstance(allergy_filters, dict) and allergy_filters.get("active"):
            keywords = allergy_filters.get("keywords") or []
            if keywords:
                filters["excludeAllergens"] = [str(k) for k in keywords[:24]]
        else:
            food_allergies = _as_list(
                constraints.get("foodAllergies")
                or onboarding.get("foodAllergies")
                or nutrition_onboarding.get("foodAllergies")
            )
            legacy = _as_list(
                constraints.get("allergies") or constraints.get("allergens") or onboarding.get("allergies")
            )
            merged = food_allergies or legacy
            if merged:
                filters["excludeAllergens"] = merged
        return filters

    return filters if len(filters) > 2 else None

File: backend-node/src/lib/rag/metadataFilters.js
/**
 * Metadata-aware SQL filters for RAG retrieval (chat + catalog).
 * Filters apply during search, not only post-retrieval.
 */

const { buildAllergyFilters } = require('../plans/constraints');

const UUID_RE =
  /^[0-9a-f]{8}-[0-9a-f]{4}-[1-5][0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}$/i;

function escapeSqlLiteral(value) {
  return String(value || '').replace(/'/g, "''");
}

function asStringList(arr) {
  if (!Array.isArray(arr)) return arr != null && arr !== '' ? [String(arr).trim()].filter(Boolean) : [];
  return arr.map((v) => String(v || '').trim()).filter(Boolean);
}

function resolveFoodAllergies(constraints, onboarding) {
  const raw =
    constraints.foodAllergies ||
    constraints.allergies ||
    constraints.allergens ||
    onboarding.foodAllergies ||
    onboarding.allergies ||
    [];
  return asStringList(raw).filter((a) => a && a !== 'none');
}

function resolveReligiousDiet(constraints, onboarding) {
  const raw = constraints.religiousDiet ?? onboarding.religiousDiet;
  const list = asStringList(raw).filter((r) => r && r !== 'none');
  const dietary = list.find(
    (r) => !['ramadan', 'christian_fasting'].includes(String(r).toLowerCase()),
  );
  return dietary || list[0] || '';
}

/**
 * @param {object} [filters]
 * @returns {string} SQL AND clauses (each prefixed with AND)
 */
function buildMetadataFilterSql(filters) {
  if (!filters || typeof filters !== 'object') return '';

  const clauses = [];

  const difficulties = asStringList(filters.difficulty);
  if (difficulties.length) {
    const vals = difficulties.map((d) => `'${escapeSqlLiteral(d.toLowerCase())}'`).join(', ');
    clauses.push(
      `(LOWER(k.metadata->>'difficulty') IN (${vals}) OR k.metadata->>'difficulty' IS NULL)`
    );
  }

  const excludeDifficulties = asStringList(filters.excludeDifficulty);
  if (excludeDifficulties.length) {
    const vals = excludeDifficulties.map((d) => `'${escapeSqlLiteral(d.toLowerCase())}'`).join(', ');
    clauses.push(
      `(k.metadata->>'difficulty' IS NULL OR LOWER(k.metadata->>'difficulty') NOT IN (${vals}))`
    );
  }

  const primaryMuscles = asStringList(filters.primaryMuscles);
  if (primaryMuscles.length) {
    const muscleClauses = primaryMuscles.map((m) => {
      const lit = escapeSqlLiteral(m.toLowerCase());
      return `(
        EXISTS (
          SELECT 1 FROM jsonb_array_elements_text(COALESCE(k.metadata->'primaryMuscles', '[]'::jsonb)) AS pm(val)
          WHERE LOWER(pm.val) LIKE '%${lit}%'
        )
        OR LOWER(k.content) LIKE '%${lit}%'
        OR LOWER(k.metadata->>'name') LIKE '%${lit}%'
      )`;
    });
    clauses.push(`(${muscleClauses.join(' OR ')})`);
  }

  const excludeExerciseIds = asStringList(filters.excludeExerciseIds).filter((id) => UUID_RE.test(id));
  if (excludeExerciseIds.length) {
    const vals = excludeExerciseIds.map((id) => `'${escapeSqlLiteral(id)}'`).join(', ');
    clauses.push(`(k.metadata->>'exerciseId' IS NULL OR k.metadata->>'exerciseId' NOT IN (${vals}))`);
  }

  const dietType = String(filters.dietType || '').trim();
  if (dietType) {
    const lit = escapeSqlLiteral(dietType.toLowerCase());
    clauses.push(`(
      LOWER(k.content) LIKE '%${lit}%'
      OR LOWER(k.metadata->>'category') LIKE '%${lit}%'
      OR EXISTS (
        SELECT 1 FROM jsonb_array_elements_text(COALESCE(k.metadata->'dietTags', '[]'::jsonb)) AS dt(val)
        WHERE LOWER(dt.val) LIKE '%${lit}%'
      )
    )`);
  }

  const religiousDiet = String(filters.religiousDiet || '').trim();
  if (religiousDiet && religiousDiet !== 'none') {
    const lit = escapeSqlLiteral(religiousDiet.toLowerCase());
    clauses.push(`(
      LOWER(k.content) LIKE '%${lit}%'
      OR EXISTS (
        SELECT 1 FROM jsonb_array_elements_text(COALESCE(k.metadata->'dietTags', '[]'::jsonb)) AS rd(val)
        WHERE LOWER(rd.val) LIKE '%${lit}%'
      )
    )`);
  }

  const excludeAllergens = asStringList(filters.excludeAllergens);
  if (excludeAllergens.length) {
    for (const allergen of excludeAllergens) {
      const lit = escapeSqlLiteral(allergen.toLowerCase());
      clauses.push(`LOWER(k.content) NOT LIKE '%${lit}%'`);
    }
  }

  const docType = String(filters.docType || '').trim();
  if (docType === 'platform') {
    clauses.push(`d.level = 'L1_INTERNAL'`);
    clauses.push(`COALESCE(d.metadata->>'docType', 'platform') = 'platform'`);
  } else if (docType === 'book') {
    clauses.push(`d.level = 'L5_BOOKS'`);
  }

  const excludeTags = asStringList(filters.excludeTags);
  if (excludeTags.length) {
    for (const tag of excludeTags) {
      const lit = escapeSqlLiteral(tag);
      clauses.push(`NOT (COALESCE(d.metadata->'tags', '[]'::jsonb) ? '${lit}')`);
    }
  }

  const chunkRoles = asStringList(filters.chunkRoles);
  if (chunkRoles.length) {
    const vals = chunkRoles.map((r) => `'${escapeSqlLiteral(r)}'`).join(', ');
    clauses.push(`k.chunk_role IN (${vals})`);
  } else {
    // Default: searchable chunks only (child + standalone, not parent storage rows)
    clauses.push(`k.chunk_role IN ('child', 'standalone')`);
  }

  if (filters.requireEmbedding === true) {
    clauses.push(`k.embedding IS NOT NULL`);
  }

  const embeddingModel = String(filters.embeddingModel || '').trim();
  if (embeddingModel) {
    clauses.push(`k.embedding_model = '${escapeSqlLiteral(embeddingModel)}'`);
  }

  const embeddingVersion = String(filters.embeddingVersion || '').trim();
  if (embeddingVersion) {
    clauses.push(`k.embedding_version = '${escapeSqlLiteral(embeddingVersion)}'`);
  }

  const localeFilter = String(filters.locale || '').trim();
  if (localeFilter === 'en' || localeFilter === 'ar') {
    clauses.push(`(d.locale = '${escapeSqlLiteral(localeFilter)}' OR d.locale = 'en')`);
  }

  return clauses.length ? clauses.map((c) => `AND ${c}`).join('\n      ') : '';
}

/**
 * Build metadata filters from chat intent + CAG bundle (mirrors ai-service metadata_filters.py).
 * @param {{ intent?: string, contextBundle?: object, locale?: string }} opts
 * @returns {object|null}
 */
function buildChatMetadataFilters({ intent, contextBundle, locale } = {}) {
  const bundle = contextBundle || {};
  const constraints = bundle.constraints || {};
  const onboarding = bundle.onboardingSummary || bundle.onboardingByFlow?.nutrition || {};
  const profile = bundle.profile || {};

  const filters = {
    chunkRoles: ['child', 'standalone'],
    requireEmbedding: true,
  };

  const resolvedIntent = String(intent || 'general');

  if (resolvedIntent === 'platform_help' || resolvedIntent === 'unclear') {
    filters.docType = 'platform';
    filters.excludeTags = ['catalog', 'books'];
    if (locale === 'en' || locale === 'ar') {
      filters.locale = locale;
    }
    return filters;
  }

  if (resolvedIntent === 'exercise_alternative' || resolvedIntent === 'workout') {
    const level = String(profile.fitnessLevel || onboarding.fitnessLevel || '').toLowerCase();
    if (level.includes('beginner') || level.includes('novice')) {
      filters.difficulty = ['beginner'];
      filters.excludeDifficulty = ['advanced'];
    } else if (level.includes('advanced') || level.includes('expert')) {
      filters.difficulty = ['intermediate', 'advanced'];
    } else if (level) {
      filters.difficulty = ['beginner', 'intermediate'];
    }

    const injuries = Array.isArray(constraints.injuries)
      ? constraints.injuries.filter((i) => i && i !== 'none')
      : [];
    if (injuries.length) {
      filters.excludeExerciseNames = injuries.map(String);
    }

    const workoutToday = bundle.workoutToday || {};
    const exercises = workoutToday.exercises || workoutToday.loggedExercises || [];
    if (Array.isArray(exercises) && exercises.length) {
      const first = exercises[0];
      const muscles =
        first?.primaryMuscles ||
        first?.muscles ||
        (first?.muscleGroup ? [first.muscleGroup] : []);
      if (Array.isArray(muscles) && muscles.length) {
        filters.primaryMuscles = muscles.map(String);
      }
    }
    return filters;
  }

  if (resolvedIntent === 'nutrition') {
    const dietType =
      constraints.dietType ||
      onboarding.dietType ||
      bundle.onboardingByFlow?.nutrition?.dietType;
    if (dietType) filters.dietType = String(dietType);

    const religiousDiet = resolveReligiousDiet(constraints, onboarding);
    if (religiousDiet) filters.religiousDiet = String(religiousDiet);

    const allergyOnboarding = {
      foodAllergies: resolveFoodAllergies(constraints, onboarding),
      foodAllergiesOther:
        constraints.foodAllergiesOther ||
        onboarding.foodAllergiesOther ||
        onboarding.allergiesOther ||
        null,
    };
    const allergyFilters = buildAllergyFilters(allergyOnboarding);
    if (allergyFilters.active) {
      filters.excludeAllergens = allergyFilters.keywords.slice(0, 24);
      filters.allergyCodes = allergyFilters.codes;
    }
    return filters;
  }

  return Object.keys(filters).length > 2 ? filters : null;
}

module.exports = {
  buildMetadataFilterSql,
  buildChatMetadataFilters,
  escapeSqlLiteral,
};
